src/data/setoid.lean

```diff
@@ -0,0 +1,580 @@
+  rw Sup_eq_eqv_gen, 
+  congr,
+  ext x y,
+  erw [Sup_image, supr_apply, supr_apply, supr_Prop_eq], 
+  split,
+    rintro ⟨r, h, hr⟩, 
+    use r,
+    rw [supr_apply, supr_apply, supr_Prop_eq], 
+    exact ⟨h, hr⟩,
+  rintro ⟨r, h⟩, 
+  obtain ⟨h, hr⟩ := by rw [supr_apply, supr_apply, supr_Prop_eq] at h; exact h,
+  exact ⟨r, h, hr⟩,
+end 
+
+/-- The equivalence closure of an equivalence relation r is r. -/
+@[simp] lemma eqv_gen_of_setoid (r : setoid α) : eqv_gen.setoid r.r = r :=
+le_antisymm (by rw eqv_gen_eq; exact Inf_le _ r (λ _ _ h, h)) eqv_gen.rel
+
+/-- Equivalence closure is idempotent. -/
+@[simp] lemma eqv_gen_idem (r : α → α → Prop) : 
+  eqv_gen.setoid (eqv_gen.setoid r).rel = eqv_gen.setoid r :=
+eqv_gen_of_setoid _ 
+
+/-- The equivalence closure of a binary relation r is contained in any equivalence
+    relation containing r. -/
+theorem eqv_gen_le {r : α → α → Prop} {s : setoid α} (h : ∀ x y, r x y → s.rel x y) : 
+  eqv_gen.setoid r ≤ s :=
+by rw eqv_gen_eq; exact Inf_le _ _ h
+
+/-- Equivalence closure of binary relations is monotonic. -/
+theorem eqv_gen_mono {r s : α → α → Prop} (h : ∀ x y, r x y → s x y) : 
+  eqv_gen.setoid r ≤ eqv_gen.setoid s :=
+eqv_gen_le $ λ _ _ hr, eqv_gen.rel _ _ $ h _ _ hr
+
+/-- There is a Galois insertion of equivalence relations on α into binary relations 
+    on α, with equivalence closure the lower adjoint. -/
+def gi : @galois_insertion (α → α → Prop) (setoid α) _ _ eqv_gen.setoid rel :=
+{ choice := λ r h, eqv_gen.setoid r,
+  gc := λ r s, ⟨λ H _ _ h, H _ _ $ eqv_gen.rel _ _ h, λ H, eqv_gen_of_setoid s ▸ eqv_gen_mono H⟩,
+  le_l_u := λ x, (eqv_gen_of_setoid x).symm ▸ le_refl x,
+  choice_eq := λ _ _, rfl}
+
+open function
+
+/-- A function from α to β is injective iff its kernel is the bottom element of the complete lattice
+    of equivalence relations on α. -/
+theorem injective_iff_ker_bot (f : α → β) :
+  injective f ↔ ker f = ⊥ := 
+⟨λ hf, setoid.ext' $ λ x y, ⟨λ h, hf h, λ h, h ▸ rfl⟩, 
+ λ hk x y h, show rel ⊥ x y, from hk ▸ (show (ker f).rel x y, from h)⟩
+
+/-- The elements related to x ∈ α by the kernel of f are those in the preimage of f(x) under f. -/
+lemma ker_apply_eq_preimage (f : α → β) (x) : (ker f).rel x = f ⁻¹' {f x} :=
+set.ext $ λ x,
+  ⟨λ h, set.mem_preimage.2 (set.mem_singleton_iff.2 h.symm),
+   λ h, (set.mem_singleton_iff.1 (set.mem_preimage.1 h)).symm⟩
+
+/-- The uniqueness part of the universal property for quotients of an arbitrary type. -/
+theorem lift_unique {r : setoid α} {f : α → β} (H : r ≤ ker f) (g : quotient r → β)
+  (Hg : f = g ∘ quotient.mk) : quotient.lift f H = g :=
+by ext; rcases x; erw [quotient.lift_beta f H, Hg]; refl
+
+/-- Given a map f from α to β, the natural map from the quotient of α by the kernel of f is 
+    injective. -/
+lemma injective_ker_lift (f : α → β) : injective (@quotient.lift _ _ (ker f) f (λ _ _ h, h)) :=
+λ x y, by apply quotient.induction_on₂' x y; intros a b; 
+  show (ker f).rel a b → _; from quotient.sound'
```

Comment:
This looks like it could just as well be a term-mode proof

Reply:
Indeed, the elaborator is being a little uncooperative, but try
```lean
λ x y, quotient.induction_on₂' x y $ λ a b h, quotient.sound' h
```